### ISSUE TITLE
Problem with deserializing a model with custom pipeline (new in v3, due to the new add_pipe paradigm and factories)

### ISSUE THREAD

**Issue by nrodnova** [explosion/spaCy#7696]
## How to reproduce the behavior

This is a pretty involved topic, please bear with me. Essentially, new `add_pipe` signature forces all the parameters that are passed to the pipe's constructor be in `config` variable, which gets serialized into `config.cfg` (say, `EntityRuler`'s patterns). Then, the same parameters get serialized again in each pipe's directory in `to_disk` calls. When deserializing the model, stuff from `config.cfg` gets passed to the factory calls and subsequently to the constructors, and then the same data is populated in `from_disk`. I will give an example that also shows a problem/bug in `EntityRuler`.

I have a large project with multiple custom pipes, and the way it works is that I create a custom model, based on a pre-packaged model, and then save it using `to_disk`. Later, in a separate project, I  load it using `spacy.load` and use as a regular spacy model.

```
import spacy
from spacy.language import Language
from spacy.pipeline import EntityRuler

@Language.factory('custom_ner')
def create_entity_ruler(nlp, name, cfg) :
    entity_ruler = EntityRuler(nlp, **cfg)
    entity_ruler.name = name
    return entity_ruler

# In real life this is a huge function that creates multiple pipes
# model_parameters is a dictionary of dictionaries, with keys corresponding to pipe names,
# the way it was encouraged in spacy 2
def create_model(language, model_parameters) :
    nlp = spacy.load(language, exclude = ['parser', 'ner', 'attribute_ruler', 'lemmatizer'])
    for name, pipe_config in model_parameters.items() :
        print("Adding pipe: %s" % name)
        nlp.add_pipe(name, config = {'cfg' : pipe_config})
    return nlp


patterns = [
    {'pattern' : [{'LOWER' : {'IN' : ['right', 'left']}}, {'LOWER' : 'corner'}], 'label' : 'LOCATION', 'id' : 'corner'},
    {'pattern' : [{'LOWER' : 'left'}], 'label' : 'LOCATION', 'id' : ''},
    {'pattern' : [{'LOWER' : 'right'}], 'label' : 'LOCATION', 'id' : ''},  
]
model_parameters = {'custom_ner' : {'patterns' : patterns}}

nlp = create_model('en_core_web_sm', model_parameters)

# Print EntityRuler's matcher patterns
matcher_patterns = nlp.get_pipe('custom_ner').matcher._patterns
for label_hash, value in matcher_patterns.items() :
    print(nlp.vocab.strings[label_hash], value)
```
Output at this point (patterns of the `EntityRuler`'s matcher):
```
LOCATION||corner [[{'LOWER': {'IN': ['right', 'left']}}, {'LOWER': 'corner'}]]
LOCATION|| [[{'LOWER': 'left'}], [{'LOWER': 'right'}]]
```
Now, save and reload the model:
```
file_name = ensure_path('test_model')
nlp.to_disk(file_name)
```
For the `custom_ner` pipe, the rules got saved in `config.cfg` and also in `custom_ner` directory in `patterns.jsonl`
```
# Load new model from the same file
nlp1 = spacy.load(file_name)

#Print patterns:
matcher_patterns = nlp1.get_pipe('custom_ner').matcher._patterns
for label_hash, value in matcher_patterns.items() :
    print(nlp1.vocab.strings[label_hash], value)
```
Output after loading the model:
```
LOCATION||corner [[{'LOWER': {'IN': ['right', 'left']}}, {'LOWER': 'corner'}], [{'LOWER': {'IN': ['right', 'left']}}, {'LOWER': 'corner'}]]
LOCATION|| [[{'LOWER': 'left'}], [{'LOWER': 'right'}]]
LOCATION [[{'LOWER': 'left'}], [{'LOWER': 'right'}]]
```
Now, there are 2 instances of each rule in the matcher. On top of that, the patterns with explicitly specified `id == ''`  are treated differently in `__init__` and in `from_disk`of `EntityRuler`. See, there are 2 labels now for the duplicated rule : `LOCATION` and `LOCATION||`

In my real system, I am also getting my entities sometimes labeled with `LOCATION||`, in what seems to be a random fashion, but I cannot reproduce that one with this small example.

So, in this easy example, to avoid all the problems mentioned above, I have to create an empty `EntityRuler` in my `create_model` function, call `add_pipe`, then get the pipe and populate the rules and other parameters using `set` methods. Which seems wrong and excessive. In more complicated cases, I do need to pass some data to the pipe's constructors without spacy blindly serializing it in `config.cfg`. Isn't `to_disk` meant to be a way to control what gets serialized and what doesn't?

## Info about spaCy

- **spaCy version:** 3.0.5
- **Platform:** Darwin-18.7.0-x86_64-i386-64bit
- **Python version:** 3.7.3
- **Pipelines:** en_core_web_lg (3.0.0), en_core_web_md (3.0.0), en_core_web_sm (3.0.0), en_core_web_trf (3.0.0)
-

**Comment by adrianeboyd** [comment]
Hi, this is indeed something that's quite different in v3, and it's something that we struggled with a bit ourselves when designing the config system, because we ran into similar issues with vectors and other data that was loaded from an external source once and then serialized with the pipeline.

Our solution was to move this kind of data to the `[initialize]` block of the config. For each component `component.initialize()` is run once when the model is initialized, and then from that point on the data is serialized with the pipeline and doesn't need to be loaded again. The `[initialize]` block in the config records how the components were initialized so that the pipeline setup is easy to reproduce, but users loading the pipeline in the future don't need to have these external sources available. Since you typically don't want the patterns themselves in the config, they can be loaded using a registered function rather than being included in the config itself.

The `[initialize]` part of the config would look like the example in the sidebar here: https://spacy.io/api/entityruler#initialize

If you'll notice, the `entity_ruler` component does not include the patterns as part of its component config (even though the constructor API still includes the `patterns` option as a holdover from v2):

https://github.com/explosion/spaCy/blob/8008e2f75b93505734018a072e4131650459ec3a/spacy/pipeline/entityruler.py#L37-L52

The initialization and serialization will become much easier if your custom NER components have a similar design that moves the patterns and other data like this into `initialize` with registered readers to load the data. The registered functions need to be available on init (with `spacy train --code` if you have custom functions), but like the external sources, they also don't need to be included as part of the final model package because they're only run once when the pipeline is initialized right before training.
